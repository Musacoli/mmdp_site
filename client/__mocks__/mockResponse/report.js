export const reportsResponse = {
  response: {
    reports: [
      {
        _id: '5c5afa15bc8f70175006ff02',
        updatedAt: '2019-02-06T15:31:38.014Z',
        createdAt: '2019-02-06T15:15:38.874Z',
        reportType: 'annual',
        title: 'Edited awesome report 1',
        archived: false,
        reportFile: {},
      },
      {
        _id: '5c5afb478e9ced17f8d59c5b',
        updatedAt: '2019-02-10T01:40:01.168Z',
        createdAt: '2019-02-06T15:20:44.723Z',
        reportType: 'annual',
        title: 'Awesome report 1',
        archived: false,
        reportFile: {},
      },
      {
        _id: '5c603a081e36fa24d709a642',
        updatedAt: '2019-02-10T14:49:46.720Z',
        createdAt: '2019-02-10T14:49:46.720Z',
        reportType: 'quarterly',
        title: 'June suprising fact',
        archived: true,
        reportFile: {},
      },
      {
        _id: '5c6042d991e77c29420808ff',
        updatedAt: '2019-02-10T15:27:23.314Z',
        createdAt: '2019-02-10T15:27:23.314Z',
        reportType: 'quarterly',
        title: 'A new dawn',
        archived: true,
        reportFile: {},
      },
      {
        _id: '5c604d89ad1d332da76a46ed',
        updatedAt: '2019-02-10T16:13:01.885Z',
        createdAt: '2019-02-10T16:13:01.885Z',
        reportType: 'annual',
        title: 'We made it',
        archived: false,
        reportFile: {},
      },
      {
        _id: '5c604e01ad1d332da76a46ef',
        updatedAt: '2019-02-10T16:14:59.646Z',
        createdAt: '2019-02-10T16:14:59.646Z',
        reportType: 'annual',
        title: 'Repo',
        archived: true,
        reportFile: {},
      },
      {
        _id: '5c604e83ad1d332da76a46f0',
        updatedAt: '2019-02-10T16:17:11.737Z',
        createdAt: '2019-02-10T16:17:11.737Z',
        reportType: 'annual',
        title: 'Amazing year',
        archived: true,
        reportFile: {},
      },
      {
        _id: '5c605087ad1d332da76a46f1',
        updatedAt: '2019-02-10T16:25:47.295Z',
        createdAt: '2019-02-10T16:25:47.295Z',
        reportType: 'quarterly',
        title: 'Amazing quarter',
        archived: false,
        reportFile: {},
      },
      {
        _id: '5c6050aaad1d332da76a46f2',
        updatedAt: '2019-02-10T16:26:21.179Z',
        createdAt: '2019-02-10T16:26:21.179Z',
        reportType: 'annual',
        title: 'A new dawn',
        __v: 0,
        archived: true,
        reportFile: {},
      },
    ],
    meta: {
      currentPage: 1,
      perPage: 20,
      totalPages: 1,
      total: 9,
    },
  },
};

export const archivedResponse = {
  report: {
    ...reportsResponse.response.reports[0],
    archived: true,
  },
};
