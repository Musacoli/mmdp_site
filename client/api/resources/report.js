import axios from 'axios';
import baseURL from '..';

const url = `${baseURL}/api/v1`;
// eslint-disable-next-line import/prefer-default-export
export const createReport = data => axios.post(`${url}/resources/report`, data);
