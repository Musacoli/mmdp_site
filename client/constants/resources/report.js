export const ADD_REPORT = 'ADD_REPORT';
export const ADD_REPORT_SUCCESS = 'ADD_REPORT_SUCCESS';
export const ADD_REPORT_FAILURE = 'ADD_REPORT_FAILURE';
export const DELETE_REPORT = 'DELETE_REPORT';
export const DELETE_REPORT_SUCCESS = 'DELETE_REPORT_SUCCESS';
export const DELETE_REPORT_FAILURE = 'DELETE_REPORT_FAILURE';
export const EDIT_REPORT = 'EDIT_REPORT';
export const EDIT_REPORT_SUCCESS = 'EDIT_REPORT_SUCCESS';
export const EDIT_REPORT_FAILURE = 'EDIT_REPORT_FAILURE';
export const FETCH_REPORT = 'FETCH_REPORT';
export const FETCH_REPORT_SUCCESS = 'FETCH_REPORT_SUCCESS';
export const FETCH_REPORT_FAILURE = 'FETCH_REPORT_FAILURE';
export const FETCH_REPORTS = 'FETCH_REPORTS';
export const FETCH_REPORTS_SUCCESS = 'FETCH_REPORTS_SUCCESS';
export const FETCH_REPORTS_FAILURE = 'FETCH_REPORTS_FAILURE';
export const ARCHIVE_REPORT = 'ARCHIVE_REPORT';
export const ARCHIVE_REPORT_SUCCESS = 'ARCHIVE_REPORT_SUCCESS';
export const ARCHIVE_REPORT_FAILURE = 'ARCHIVE_REPORT_FAILURE';
