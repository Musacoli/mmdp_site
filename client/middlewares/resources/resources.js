import { put, call, takeLatest } from 'redux-saga/effects';
import toastr from 'toastr';
import axios from 'axios';
import { ADD_USER_RESEARCH } from '../../constants/resources/research';
import { addResearchSuccessfull } from '../../store/actions/resources/research';

toastr.options = {
  positionClass: 'toast-top-center',
  preventDuplicates: true,
};
const apiRequest = (url, data) => axios
  .post(url, data)
  .then(response => response.data)
  .catch((error) => {
    if (error.response) {
      const response = error.response.data;
      return response;
    }
  });

export function* addUserResearch(action) {
  try {
    const RESEARCH_URL = 'http://localhost:3000/api/v1/resources/research';
    const USERDETAILS = action.payload;
    const response = yield call(apiRequest, RESEARCH_URL, USERDETAILS);
    yield put(addResearchSuccessfull(response));
    if (response.status === 'success') {
      toastr.success(response.message);
    } else {
      toastr.warning(response.message);
    }
  } catch (error) {
    throw error;
  }
}

export default function* addUserResearchWatcher() {
  yield takeLatest(ADD_USER_RESEARCH, addUserResearch);
}
