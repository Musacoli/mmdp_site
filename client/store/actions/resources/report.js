import {
  ADD_REPORT,
  ADD_REPORT_SUCCESS,
  ADD_REPORT_FAILURE,
} from '../../../constants/resources/report';

export const addReport = (payload) => ({
  type: ADD_REPORT,
  payload,
});

export const addReportSuccessful = (payload) => ({
  type: ADD_REPORT_SUCCESS,
  payload,
});

export const addReportFailure = (payload) => ({
  type: ADD_REPORT_FAILURE,
  payload,
});
