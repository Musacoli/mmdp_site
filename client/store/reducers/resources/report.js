import {
  ADD_REPORT,
  ADD_REPORT_SUCCESS,
  ADD_REPORT_FAILURE,
  DELETE_REPORT,
  DELETE_REPORT_SUCCESS,
  DELETE_REPORT_FAILURE,
  EDIT_REPORT,
  EDIT_REPORT_SUCCESS,
  EDIT_REPORT_FAILURE,
  FETCH_REPORT,
  FETCH_REPORT_SUCCESS,
  FETCH_REPORT_FAILURE,
  FETCH_REPORTS,
  FETCH_REPORTS_SUCCESS,
  FETCH_REPORTS_FAILURE,
  ARCHIVE_REPORT,
  ARCHIVE_REPORT_SUCCESS,
  ARCHIVE_REPORT_FAILURE,
} from '../../../constants/resources/report';

export const initialState = {
  loading: false,
  response: {},
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case ADD_REPORT:
    case EDIT_REPORT:
    case DELETE_REPORT:
    case ARCHIVE_REPORT:
    case FETCH_REPORTS:
      return { ...state, loading: true, ...payload };
    case FETCH_REPORT:
      return { ...state, loading: true };
    case ADD_REPORT_SUCCESS:
    case EDIT_REPORT_SUCCESS:
    case FETCH_REPORT_SUCCESS:
    case FETCH_REPORTS_SUCCESS:
      return { ...state, loading: false, ...payload };
    case ARCHIVE_REPORT_SUCCESS: {
      const { id: archivedReportId, action, ...restState } = state;
      const { report, ...restResponse } = payload.response;
      const reportIndex = state.response.reports.findIndex(
        ({ _id: id }) => id === archivedReportId,
      );
      if (reportIndex > -1) restState.response.reports[reportIndex] = report;
      const response = {
        ...restResponse,
        reports: restState.response.reports,
      };
      return { ...restState, loading: false, response };
    }
    case DELETE_REPORT_SUCCESS: {
      const { id: deletedReportId, ...restState } = state;
      const response = {
        ...payload.response,
        reports: state.response.reports.filter(
          ({ _id: id }) => id !== deletedReportId,
        ),
      };
      return { ...restState, loading: false, response };
    }
    case ADD_REPORT_FAILURE:
    case EDIT_REPORT_FAILURE:
    case DELETE_REPORT_FAILURE:
    case FETCH_REPORT_FAILURE:
    case FETCH_REPORTS_FAILURE:
    case ARCHIVE_REPORT_FAILURE:
      return { ...state, loading: false, ...payload };
    default:
      return state;
  }
};
