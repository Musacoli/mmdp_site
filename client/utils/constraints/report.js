const reportFormConstraint = {
  title: {
    presence: {
      allowEmpty: false,
      message: '^Enter the report title',
    },
    length: {
      minimum: 4,
      message: 'must be at least 4 characters',
    },
  },
  reportFile: {
    presence: {
      allowEmpty: false,
      message: '^Select a report file to upload',
    },
  },
  reportType: {
    presence: {
      allowEmpty: false,
      message: '^Select a report type',
    },
    inclusion: {
      within: ['quarterly', 'annual'],
      message:
        '^Report type can only be a quarterly report or an annual report',
    },
  },
};

export default reportFormConstraint;
