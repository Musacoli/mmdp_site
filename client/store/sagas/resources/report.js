import { delay } from 'redux-saga';
import { put, call, takeLatest } from 'redux-saga/effects';
import toastr from 'toastr';
import {
  ADD_REPORT,
  EDIT_REPORT,
  DELETE_REPORT,
  FETCH_REPORT,
  FETCH_REPORTS,
  ARCHIVE_REPORT,
} from '../../../constants/resources/report';
import * as actions from '../../actions/resources/report';
import * as request from '../../../utils/resources/report';
import { deriveError } from '../../../utils/helper';

export function* sendReport({ payload }) {
  try {
    const { id, formData, mode } = payload;
    let response;
    let successAction;
    if (mode === 'edit') {
      response = yield call(request.updateReport, id, formData);
      successAction = actions.editReportSuccessful;
    } else {
      response = yield call(request.createReport, formData);
      successAction = actions.addReportSuccessful;
    }
    toastr.success(response.data.message);
    yield delay(500);
    const res = {
      statusCode: response.status,
      report: response.data.data.report,
    };
    yield put(successAction({ response: res }));
  } catch (err) {
    const error = deriveError(err);
    toastr.warning(error);
    yield put(actions.addReportFailure({ error }));
  }
}

export function* fetchReport({ payload }) {
  try {
    const res = yield call(request.fetchReport, payload);
    const response = {
      statusCode: res.status,
      status: res.data.status,
      report: res.data.data.report,
    };
    yield put(actions.fetchReportSuccessful({ response }));
  } catch (err) {
    const error = deriveError(err);
    toastr.warning(error);
    yield put(actions.fetchReportFailure(error));
  }
}

export function* fetchReports({ payload }) {
  try {
    const res = yield call(request.fetchReports, payload);
    const response = {
      statusCode: res.status,
      status: res.data.status,
      reports: res.data.data.reports,
      pagination: res.data.data.pagination,
    };
    yield delay(1000);
    yield put(actions.fetchReportsSuccessful({ response }));
  } catch (err) {
    const error = deriveError(err);
    toastr.warning(error);
    yield put(actions.fetchReportsFailure(error));
  }
}

export function* deleteReport({ payload }) {
  try {
    const { id } = payload;
    const res = yield call(request.deleteReport, id);
    toastr.success(res.data.message);
    const response = {
      statusCode: res.status,
      status: res.data.status,
    };
    yield put(actions.deleteReportSuccessful({ response }));
  } catch (err) {
    const error = deriveError(err);
    toastr.warning(error);
    yield put(actions.deleteReportFailure(error));
  }
}

export function* archiveReport({ payload }) {
  try {
    const { id, action } = payload;
    const res = yield call(request.archiveReport, id, action);
    toastr.success(res.data.message);
    const response = {
      statusCode: res.status,
      status: res.data.status,
      report: res.data.data.report,
    };
    yield put(actions.archiveReportSuccessful({ response }));
  } catch (err) {
    const error = deriveError(err);
    toastr.warning(error);
    yield put(actions.archiveReportFailure(error));
  }
}

export function* watchReport() {
  yield takeLatest(ADD_REPORT, sendReport);
  yield takeLatest(EDIT_REPORT, sendReport);
  yield takeLatest(DELETE_REPORT, deleteReport);
  yield takeLatest(FETCH_REPORT, fetchReport);
  yield takeLatest(FETCH_REPORTS, fetchReports);
  yield takeLatest(ARCHIVE_REPORT, archiveReport);
}
