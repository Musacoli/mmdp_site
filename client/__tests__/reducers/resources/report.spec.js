import reportReducer, {
  initialState,
} from '../../../store/reducers/resources/report';
import * as actions from '../../../store/actions/resources/report';
import {
  archivedResponse,
  reportsResponse,
} from '../../../__mocks__/mockResponse/report';

const loadingState = { loading: true, response: {} };
const notLoadingState = { loading: false, response: {} };
const itemToModify = '5c5afa15bc8f70175006ff02';

describe('Report reducer', () => {
  it('should return the initial state when action type is not known', () => {
    const action = { type: 'UNKNOWN_ACTION' };
    const state = reportReducer(undefined, action);
    expect(state).toEqual(initialState);
  });
  it('should return handle ADD_REPORT action correctly', () => {
    const state = reportReducer(undefined, actions.addReport());
    expect(state).toEqual(loadingState);
  });
  it('should return handle ADD_REPORT_SUCCESS action correctly', () => {
    const state = reportReducer(loadingState, actions.addReportSuccessful());
    expect(state).toEqual(notLoadingState);
  });
  it('should return handle ADD_REPORT_FAILURE action correctly', () => {
    const state = reportReducer(loadingState, actions.addReportFailure());
    expect(state).toEqual(notLoadingState);
  });
  it('should return handle FETCH_REPORT action correctly', () => {
    const state = reportReducer(notLoadingState, actions.fetchReport());
    expect(state).toEqual(loadingState);
  });
  it('should return handle ARCHIVE_REPORT_SUCCESS action correctly', () => {
    const state = reportReducer(
      {
        id: itemToModify,
        action: 'archive',
        loading: false,
        ...reportsResponse,
      },
      actions.archiveReportSuccessful({
        response: archivedResponse,
      }),
    );
    const archivedReport = state.response.reports.find(
      ({ _id: id }) => id === itemToModify,
    );
    expect(archivedReport.archived).toEqual(true);
    expect(state).toMatchSnapshot();
  });
  it('should return handle DELETE_REPORT_SUCCESS action correctly', () => {
    const state = reportReducer(
      {
        id: itemToModify,
        action: 'delete',
        loading: false,
        ...reportsResponse,
      },
      actions.deleteReportSuccessful({
        response: archivedResponse,
      }),
    );
    const archivedReport = state.response.reports.find(
      ({ _id: id }) => id === itemToModify,
    );
    expect(archivedReport).toBe(undefined);
    expect(state).toMatchSnapshot();
  });
});
