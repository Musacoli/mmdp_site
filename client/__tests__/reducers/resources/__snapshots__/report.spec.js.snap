// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report reducer should return handle ARCHIVE_REPORT_SUCCESS action correctly 1`] = `
Object {
  "loading": false,
  "response": Object {
    "reports": Array [
      Object {
        "_id": "5c5afa15bc8f70175006ff02",
        "archived": true,
        "createdAt": "2019-02-06T15:15:38.874Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Edited awesome report 1",
        "updatedAt": "2019-02-06T15:31:38.014Z",
      },
      Object {
        "_id": "5c5afb478e9ced17f8d59c5b",
        "archived": false,
        "createdAt": "2019-02-06T15:20:44.723Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Awesome report 1",
        "updatedAt": "2019-02-10T01:40:01.168Z",
      },
      Object {
        "_id": "5c603a081e36fa24d709a642",
        "archived": true,
        "createdAt": "2019-02-10T14:49:46.720Z",
        "reportFile": Object {},
        "reportType": "quarterly",
        "title": "June suprising fact",
        "updatedAt": "2019-02-10T14:49:46.720Z",
      },
      Object {
        "_id": "5c6042d991e77c29420808ff",
        "archived": true,
        "createdAt": "2019-02-10T15:27:23.314Z",
        "reportFile": Object {},
        "reportType": "quarterly",
        "title": "A new dawn",
        "updatedAt": "2019-02-10T15:27:23.314Z",
      },
      Object {
        "_id": "5c604d89ad1d332da76a46ed",
        "archived": false,
        "createdAt": "2019-02-10T16:13:01.885Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "We made it",
        "updatedAt": "2019-02-10T16:13:01.885Z",
      },
      Object {
        "_id": "5c604e01ad1d332da76a46ef",
        "archived": true,
        "createdAt": "2019-02-10T16:14:59.646Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Repo",
        "updatedAt": "2019-02-10T16:14:59.646Z",
      },
      Object {
        "_id": "5c604e83ad1d332da76a46f0",
        "archived": true,
        "createdAt": "2019-02-10T16:17:11.737Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Amazing year",
        "updatedAt": "2019-02-10T16:17:11.737Z",
      },
      Object {
        "_id": "5c605087ad1d332da76a46f1",
        "archived": false,
        "createdAt": "2019-02-10T16:25:47.295Z",
        "reportFile": Object {},
        "reportType": "quarterly",
        "title": "Amazing quarter",
        "updatedAt": "2019-02-10T16:25:47.295Z",
      },
      Object {
        "__v": 0,
        "_id": "5c6050aaad1d332da76a46f2",
        "archived": true,
        "createdAt": "2019-02-10T16:26:21.179Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "A new dawn",
        "updatedAt": "2019-02-10T16:26:21.179Z",
      },
    ],
  },
}
`;

exports[`Report reducer should return handle DELETE_REPORT_SUCCESS action correctly 1`] = `
Object {
  "action": "delete",
  "loading": false,
  "response": Object {
    "report": Object {
      "_id": "5c5afa15bc8f70175006ff02",
      "archived": true,
      "createdAt": "2019-02-06T15:15:38.874Z",
      "reportFile": Object {},
      "reportType": "annual",
      "title": "Edited awesome report 1",
      "updatedAt": "2019-02-06T15:31:38.014Z",
    },
    "reports": Array [
      Object {
        "_id": "5c5afb478e9ced17f8d59c5b",
        "archived": false,
        "createdAt": "2019-02-06T15:20:44.723Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Awesome report 1",
        "updatedAt": "2019-02-10T01:40:01.168Z",
      },
      Object {
        "_id": "5c603a081e36fa24d709a642",
        "archived": true,
        "createdAt": "2019-02-10T14:49:46.720Z",
        "reportFile": Object {},
        "reportType": "quarterly",
        "title": "June suprising fact",
        "updatedAt": "2019-02-10T14:49:46.720Z",
      },
      Object {
        "_id": "5c6042d991e77c29420808ff",
        "archived": true,
        "createdAt": "2019-02-10T15:27:23.314Z",
        "reportFile": Object {},
        "reportType": "quarterly",
        "title": "A new dawn",
        "updatedAt": "2019-02-10T15:27:23.314Z",
      },
      Object {
        "_id": "5c604d89ad1d332da76a46ed",
        "archived": false,
        "createdAt": "2019-02-10T16:13:01.885Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "We made it",
        "updatedAt": "2019-02-10T16:13:01.885Z",
      },
      Object {
        "_id": "5c604e01ad1d332da76a46ef",
        "archived": true,
        "createdAt": "2019-02-10T16:14:59.646Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Repo",
        "updatedAt": "2019-02-10T16:14:59.646Z",
      },
      Object {
        "_id": "5c604e83ad1d332da76a46f0",
        "archived": true,
        "createdAt": "2019-02-10T16:17:11.737Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "Amazing year",
        "updatedAt": "2019-02-10T16:17:11.737Z",
      },
      Object {
        "_id": "5c605087ad1d332da76a46f1",
        "archived": false,
        "createdAt": "2019-02-10T16:25:47.295Z",
        "reportFile": Object {},
        "reportType": "quarterly",
        "title": "Amazing quarter",
        "updatedAt": "2019-02-10T16:25:47.295Z",
      },
      Object {
        "__v": 0,
        "_id": "5c6050aaad1d332da76a46f2",
        "archived": true,
        "createdAt": "2019-02-10T16:26:21.179Z",
        "reportFile": Object {},
        "reportType": "annual",
        "title": "A new dawn",
        "updatedAt": "2019-02-10T16:26:21.179Z",
      },
    ],
  },
}
`;
